src/testcafe-runner.js

```diff
@@ -96,33 +97,36 @@ async function run (runCfgPath, suiteName) {
     });
 
     let endTime = new Date().toISOString();
+    passed = results === 0;
 
-    // Retain the assets now
-    if (process.env.SAUCE_USERNAME && process.env.SAUCE_ACCESS_KEY && !process.env.SAUCE_VM) {
-      console.log(`Reporting assets in '${assetsPath}' to Sauce Labs`);
-      await sauceReporter({
-        browserName,
-        assetsPath,
-        results,
-        assets: [
-          path.join(assetsPath, 'report.xml'),
-          path.join(assetsPath, 'report.json'),
-          path.join(assetsPath, 'video.mp4'),
-          path.join(assetsPath, 'console.log'),
-        ],
-        startTime,
-        endTime,
-      });
-    } else if (!process.env.SAUCE_VM) {
+    if (process.env.SAUCE_VM) {
+      return passed;
+    }
+    if (!process.env.SAUCE_USERNAME && !process.env.SAUCE_ACCESS_KEY) {
       console.log('Skipping asset uploads! Remember to setup your SAUCE_USERNAME/SAUCE_ACCESS_KEY');
+      return passed;
     }
-    passed = results === 0;
+
+    console.log(`Reporting assets in '${assetsPath}' to Sauce Labs`);
+    const reported = await sauceReporter({
+      browserName,
+      assetsPath,
+      results,
+      assets: [
+        path.join(assetsPath, 'report.xml'),
+        path.join(assetsPath, 'report.json'),
+        path.join(assetsPath, 'video.mp4'),
+        path.join(assetsPath, 'console.log'),
+      ],
+      startTime,
+      endTime,
+    });
+    passed = results === 0 && reported;
```

Comment:
Hm. Just thinking out loud, but should we really fail the tests/pipeline if the fail to report? After all, the tests did pass.

Reply:
IMO it should.

For several reasons:
- Part of the testing is the reporting to have trends insight.
- If runner does not return error when failing, saucectl will not be aware of that (it can be fixed, but will involve deeper changes)

It should considered to make this configurable to the user, depending if user is concerned by reporting status.

Reply:
> Part of the testing is the reporting to have trends insight.

Yup, that's an additional feature *on top* of running the test. It's not _crucial_ though.

> If runner does not return error when failing, saucectl will not be aware of that (it can be fixed, but will involve deeper changes)

It's crucial that the runner reports _test errors_ back to saucectl at all times. That part is indisputable.

My concern is that customers have long running pipelines that pass tests, but may fail reporting (for a number of reasons). Now they'd have to re-run the pipeline, delay their progress, just for the sake of reporting the results to Sauce Labs. A report they'll likely not care about too much (for that particular instance), since everything did pass.

Reply:
@alexplischke This has been updated. Failed to report will not anymore trigger a `exitCode != 0`.
It also provides a value that `saucectl` would be able to read and access to warn user that report failed.